app/component/SummaryPage/hooks/useAutofetchQuery.js

```diff
@@ -0,0 +1,238 @@
+import { useEffect, useReducer } from 'react';
+import { fetchQuery } from 'react-relay';
+
+/**
+ * Relay runtime query result re-fetch hook. Hook user must provide the set of callback functions
+ * which describe how graphql refetching should be done.
+ *
+ *
+ * @typedef {AutofetchInitializer}
+ * @property {RelayEnvironment} relayEnvironment
+ * @prop {GraphQLQuery} query
+ * @prop {Object.<String,*>} initialState
+ * @prop {Function} getPageCursor Getter function: getPageCursor(data, state) --> next query cursor string
+ * @prop {Function} onData Callback on query data received: onData(data)
+ * @prop {Function} onComplete Callback when fetch is complete and no refetching is required.
+ * @prop {Function} updateState State updater: updateState(data, state) --> new state
+ * @prop {Function} updateVariables New query variables updater: updateVariables(previousVariable, state). This is called after `updateState`.
+ * @prop {function} shouldFetchMore Predicate for refetch: shouldFetchMore(state). This is called after `updateState`.
+ *
+ * @param {AutofetchInitializer} opts
+ *
+ * @example
+ * const fetchPlan = useAutofetchQuery({
+ *   relayEnvironment,
+ *   query: PlanQuery
+ *
+ *   // initial user state
+ *   initialState: {
+ *      missingCount: 5,
+ *   },
+ *
+ *   // getter
+ *   getPageCursor: data => data.plan.previousPageCursor,
+ *
+ *   // callback
+ *   onData,
+ *
+ *   // callback
+ *   onComplete,
+ *
+ *   // update missingCount by new data
+ *   updateState: (data, state) => ({
+ *     ...state,
+ *     missingCount: clamp(state.missingCount - data.plan.itineraries.length),
+ *   }),
+ *
+ *   // update graphql query variables for refetch query
+ *   updateVariables: (variables, state) => ({
+ *     ...variables,
+ *     numItineraries: state.missingCount,
+ *   }),
+ *
+ *   // refetch condition
+ *   shouldFetchMore: state => state.missingCount > 0,
+ * })
+ *
+ * // fetch with graphql query variables
+ * fetchPlan({
+ *    time: "18:00" from: "65.1230,24.1230", to: "66.0000,35.0000"
+ *  })
+ */
+const useAutofetchQuery = ({
+  relayEnvironment,
+  query,
+  initialState,
+  getPageCursor,
+  onData,
+  onComplete,
+  updateState,
+  updateVariables,
+  shouldFetchMore,
+}) => {
+  const reducerInitState = {
+    pageCursor: undefined,
+    userState: initialState || {},
+    shouldFetch: false,
+    variables: {},
+    data: undefined,
+    status: 'READY',
+    error: undefined,
+  };
+
+  const ACTION_DATA = 'data';
+  const ACTION_START = 'start';
+  const ACTION_ERROR = 'error';
+  const ACTION_DATA_COMPLETE = 'data-complete';
+
+  const STATUS_PROCESSING = 'PROCESSING';
+  const STATUS_COMPLETE = 'COMPLETE';
+  const STATUS_FETCHING = 'FETCHING';
+  const STATUS_READY = 'READY';
```

Comment:
Should we change these to be in the enum object format and maybe in some different file we can, for example, check that the status is something like `status === AutofetchQueryStatuses.Complete`?